### PR TITLE
[FLINK-9004][tests] Implement Jepsen tests to test job availability.

### PR DESCRIPTION
**Pull request by GJL**
## What is the purpose of the change

*Use the Jepsen framework (https://github.com/jepsen-io/jepsen) to implement
tests that verify Flink's HA capabilities under real-world faults, such as
sudden TaskManager/JobManager termination, HDFS NameNode unavailability, network
partitions, etc. The Flink cluster under test is automatically deployed on YARN
(session & job mode) and Mesos.*

Previous PR got closed accidentally: https://github.com/apache/flink/pull/6239

## Brief change log

  - *Implement Jepsen tests.*

## Verifying this change

This change added tests and can be verified as follows:

  - *The changes themselves are tests.*
  - *Run Jepsen tests in docker containers.*
  - *Run unit tests with `lein test`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no** (at least not to Flink))
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** (but it will as soon as test failures appear) / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

cc: @tillrohrmann @cewood @zentol @aljoscha @yanghua

### CODE REVIEW

**Comment by GJL** [comment]
As @zentol mentioned, we might want to exclude this from create_source_release.sh.

File: jepsen-flink/.gitignore [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+.idea/
+/.lein-*
+/.nrepl-port
+/checkouts
+/classes
+/target
+pom.xml
+pom.xml.asc
+store
+bin/DataStreamAllroundTestProgram.jar
```
Comment (tillrohrmann) [inline jepsen-flink/.gitignore]: Maybe we could ignore the complete `bin/` folder. I put, for example, my `flink-dist.tgz` there and git reports it now as an untracked file.

**Comment by tillrohrmann** [comment]
Concerning @zentol's comment: We also include the `flink-end-to-end-tests` in the source release. Therefore, I think we could also include the jepsen-tests in the src release. Especially since one can execute them easily via `docker-compose`.

**Comment by zentol** [comment]
The jepsen license is not compatible with the apache license and thus must be part of the release.

**Comment by tillrohrmann** [comment]
Where exactly do you see the license conflicts @zentol? We do not include Jepsen code in our repository and since it is licensed under EPL 1.0 we should be able to link against it.

**Comment by zentol** [comment]
My information is outdated, they updated the license roughly a month ago to EPL:
https://github.com/jepsen-io/jepsen/commit/d9c8530b7a2deac2389d93639581dfa1ff0f08fc

**Comment by tillrohrmann** [comment]
I think you are right @zentol because we are linking against Jepsen `0.1.8` and Jepsen was licensed under EPL only in version `0.1.10`.

**Comment by zentol** [comment]
Note that this may also imply that the files that we add cannot be apache licensed, but I'm not completely sure about that.

**Comment by tillrohrmann** [comment]
Gary is reaching out to Apache legal to ask for advice how to handle the SOLIPSISTIC license. Apparently, we cannot easily upgrade to `0.1.10` because it contains a bug.

**Comment by GJL** [comment]
`0.1.9.` was problematic. Currently testing `0.1.10`.

**Comment by GJL** [comment]
It seems that we have to stick with `0.1.8` for now. According to legal, there are no concerns so far: https://issues.apache.org/jira/browse/LEGAL-392

File: jepsen-flink/.gitignore (hunk 1 above)
Comment (GJL) [reply jepsen-flink/.gitignore]: Good point. I fixed it.

File: flink-jepsen/src/jepsen/flink/db.clj [hunk 2]
```diff
@@ -175,7 +175,7 @@
           (c/su
             (c/exec (c/lit (str "HADOOP_CLASSPATH=`" hadoop/install-dir "/bin/hadoop classpath` "
                                 "HADOOP_CONF_DIR=" hadoop/hadoop-conf-dir
-                                " " install-dir "/bin/yarn-session.sh -d -jm 2048 -tm 2048")))
+                                " " install-dir "/bin/yarn-session.sh -d -jm 2048m -tm 2048m")))
```
Comment (GJL) [inline flink-jepsen/src/jepsen/flink/db.clj]: See https://issues.apache.org/jira/browse/FLINK-9777

**Comment by GJL** [comment]
I extended the regular `README.md`.

**Comment by tillrohrmann** [comment]
Nice, thank you. Merging this PR.

**Pull request closed without merging**